### ISSUE TITLE
Marr/Mexican hat wavelet?

### ISSUE THREAD

**Issue by redaready** [cybergeek94/img_hash#10]
it match better pictures 

http://www.phash.org/demo/

**Comment by cybergeek94** [comment]
Interesting, I've had the Harr wavelet suggested as well. However I cannot base my implementation on pHash because its license is incompatible with `img_hash`'s and could possibly cause conflicts. If you can find a high-level description of the algorithm somewhere, that would be very helpful.

**Comment by abonander** [comment]
Recently after some thought and research I've figured out how I want to implement this. The Marr wavelet is a continuous function and so needs to be discretely sampled for the transform. Clicking around relevant Wikipedia articles, I think pHash's implementation has something to do with the [Difference of Gaussians](https://en.wikipedia.org/wiki/Difference_of_Gaussians) operation which effectively performs edge detection on the image.

I plan on rewriting the API to allow the Discrete Cosine Transform as a preprocessing step for the mean, gradient and double-gradient hash types, and add Difference of Gaussians as another preprocessing step that can apply to the Blockhash algorithm as well as the others. The `image` crate provides a Gaussian blur function so all that's necessary is to implement the difference part which should be easy.

**Comment by c-edw** [comment]
You are welcome to copy my implementation of the Marr wavelet from https://github.com/c-edw/perceptual-hash-poc.

**Comment by abonander** [comment]
I'm still struggling with basic vector algebra but I'm not sure if that implementation is implementing a Marr wavelet transform. It looks like it's just multiplying the pixel values with the scaled wavelet. My uneducated guess is that what you've implemented there is just a weighted average. 

The `image` crate already provides a Gaussian blur function which I can used to implement Difference of Gaussians, though I don't actually know what its usefulness will be with the existing hash algorithms because a mostly white image with a handful of black pixels will scale down to a flat-grey image with few features. I imagine the Blockhash algorithm would be the only actually useful one to combine with DoG as a preprocessing step.

**Comment by c-edw** [comment]
The code there is a translated of the formula seen [here](https://en.wikipedia.org/wiki/Mexican_hat_wavelet).

![image](https://user-images.githubusercontent.com/38229097/50092334-ba961200-0205-11e9-9a26-2096f7e4364b.png)

It's possible that I missed something, but the algorithm seems to generate correct hashes. As an example, the code compares two images at largely different resolutions. The Hamming distance when using the Marr wavelet is zero.

**Comment by abonander** [comment]
Yes but all that does is produce the value of the wavelet at that point, then you just multiply it with the corresponding luminance values before passing it to a normal averaging hash. As far as I can tell, you've basically created a weighted mean hash that pays more attention to the center and edges and less to the intermediate areas. It's not a wavelet transform that I can see.

If you're testing with images of the exact content and aspect ratio, just different resolutions, I don't see why this hash would perform much differently than a regular mean hash.

**Comment by c-edw** [comment]
>I don't see why this hash would perform much differently than a regular mean hash.

Well, it does. The average (aHash) algorithm yields a Hamming distance of 2 in that scenario.

**Comment by abonander** [comment]
I'll try it. Like I said I haven't looked at the pHash source in order to avoid any IP conflicts since it has an incompatible license (like how ReactOS can't allow anyone to contribute who has seen Windows source code). That and it's no fun just copying someone else's implementation.

**Comment by BradKML** [comment]
If we are on the hash variant train: Take a look at these sample Python implementation https://github.com/thorn-oss/perception#supported-hashing-algorithms
